### PR TITLE
Feature/self reference

### PR DESCRIPTION
**Pull request by peterpaul**
This PR backports self-references and method invocations from zinc 0.2.x to our 0.1.5 branch.

- Self: Type alias for the surrounding struct
- self: variable reference for the surrounding struct
- method invocation: From Struct::fun(self, vars) to self.fun(vars)

### CODE REVIEW

File: zinc-compiler/src/semantic/analyzer/statement/mod.rs [hunk 1]
```diff
@@ -207,25 +224,60 @@ impl Analyzer {
...
+                }
+            }
 
-            Scope::declare_variable(
-                self.scope_stack.top(),
-                identifier,
-                ScopeVariableItem::new(is_mutable, r#type),
-            )
-            .map_err(|error| Error::Scope(identifier_location, error))?;
+            // let (identifier, is_mutable) = match argument_binding.variant {
```
Comment (peterpaul) [inline zinc-compiler/src/semantic/analyzer/statement/mod.rs:266]: Old code, can be removed

File: Cargo.lock [hunk 2]
```diff
@@ -1098,7 +1100,7 @@ dependencies = [
  "log",
  "num-bigint",
  "num-integer",
- "num-traits 0.1.43",
+ "num-traits 0.2.11",
```
Comment (mvdbos) [inline Cargo.lock:1103]: Further below it is 0.2, is that intentional? Does it make a difference?

**Review by gtillem — APPROVED** [review]
I guess you need to update the ING zinc version before you merge it.

File: zinc-compiler/src/lexical/stream/comment.rs [hunk 3]
```diff
@@ -177,15 +178,15 @@ mod tests {
...
         assert_eq!(result, expected);
     }
 
     #[test]
     fn error_not_a_comment_one_slash() {
-        let input = "/almost a comment text";
+        let input = r#"/almost a comment text"#;
         let expected = Err(Error::NotAComment);
         let result = parse(input);
         assert_eq!(result, expected);
```
[the remark this answers is not in the document]
Comment (peterpaul) [reply zinc-compiler/src/lexical/stream/comment.rs:192]: No all of these raw strings are from refactorings that I also backported.

File: Cargo.lock (hunk 2 above)
Comment (peterpaul) [reply Cargo.lock:1103]: I think that these are just some changes that I also backported. There is quite a difference between how versions were defined on the 0.1.x branch and on  the 0.2.x branch. I checked the release-notes of 'num-traits' and as first glance I didn't see any big changes or bugfixes.

**Comment by peterpaul** [comment]
> I guess you need to update the ING zinc version before you merge it.

Done! Thanks for the reminder ;)

**Review by mvdbos — APPROVED** [review]
Looks good. And tbh, I can't judge all code without doing a proper deep dive myself. I see the tests that prove it works.

**Pull request merged**